### DISCUSSION TITLE
AI Deck Issues

### DISCUSSION THREAD

**Discussion by cuber-1** [bitcraze/discussions#1467] [General]
I'm currently trying to update the AI deck firmware on my WINDOWS device following the instructions on this website: https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#setup-development-environment

I'm not able to see the message "ESP32: I (910) SYS: Initialized" in the console. Instead, this is what I see. 
SYS: ----------------------------
SYS: Crazyflie 2.1 is up and running!
SYS: Build 0:e527895e8c6e (2023.11) CLEAN
SYS: I am 0x313232313232470F001D0031 and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 0 deck(s) found
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Complementary (1) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2024.2 (CF21)
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 20496 bytes
STAB: Starting stabilizer loop
SUP: Can not fly
SUP: Ready to fly

It says 0 decks found? Could it be I attached the AI deck wrong? I thought attaching it is pretty straightforward. 

Another thing - When I was updating the firmware in the bootloader, I didn't see the message that it shows in the image of the Crazyflie service tab on the website I linked above (Writing to bcAI:esp deck memory. 

I've tried it multiple times by now but it's still not working. I'm considering trying the Gap8 bootloader fix but it looks like a need a jtag enabled programmer. But I need to know a couple things before trying this fix: If I manually flash the Gap8 bootloader with a separate Linux device, will the bootloader still work when I try to flash through my normal Windows device? 

Thanks!

**Comment by evoggy** [comment]
Hi,

The deck isn't being detected so boot loading will not work either. Do you have any decks or Crazyflies you can test with to see if it might be a hardware issue? Can you try with the latest firmware to see if there's any difference? Can you post a picture of the setup to make sure it's connected the right way?

**Reply by cuber-1** [reply]
I don't have any other decks or Crazyflies to test on. If you mean the latest firmware for the Crazyflie, I believe that is already installed. As for a picture, it's at a lab, so I'll get access to it on Monday. However, I can describe how I connected it. I took the battery deck off and insert the short end of the pins from the bottom of the AI deck to the top. The I take the long end of the pin and insert it into the top of the crazyflie, with the camera of the AI deck facing the front of the drone. I push the pin through until the AI deck is acting as a holder for the battery.

**Reply by knmcguire** [reply]
could you perhaps send a picture of how you have done your set-up? perhaps the deck has been placed the wrong way. 

Here is an tutorial how to work with expansion decks: https://www.bitcraze.io/documentation/tutorials/getting-started-with-expansion-decks/

**Reply by cuber-1** [reply]
Wow - I guess it wasn't straightforward to me after all. I flipped the AI deck over and it works now! Now I'm trying to run the Wifi streaming example, but when I try to run it it says:

C:\Users\NAME>cfloader flash aideck_gap8_wifi_img_streamer_with_ap.bin deck-bcAI:gap8-fw -w radio://0/80/2M/E7E7E7E7E7
Reset to bootloader mode ...
Failed to flash: [Errno 2] No such file or directory: 'aideck_gap8_wifi_img_streamer_with_ap.bin'

All I did was download the first link on the https://github.com/bitcraze/aideck-gap8-examples/releases website under 2022.05. Is there something else I need to do to make this work?

**Reply by cuber-1** [reply]
I actually ended up realizing that the file name I had entered was wrong - it was supposed to be  aideck_gap8_wifi_img_streamer_with_ap (2) .bin. However, now I see a different issue:

C:\Users\NAME>cfloader flash "C:\Users\NAME\aideck_gap8_wifi_img_streamer_with_ap (2).bin" deck-bcAI:gap8-fw -w radio://0/80/2M/E7E7E7E7E7
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\NAME\AppData\Local\Programs\Python\Python312\Scripts\cfloader.exe\__main__.py", line 4, in <module>
ImportError: cannot import name 'main' from 'cfloader' (C:\Users\NAME\AppData\Local\Programs\Python\Python312\Lib\site-packages\cfloader\__init__.py)

**Reply by knmcguire** [reply]
Ahh so you are running this from windows. I see that you have installed python just locally for your user and not for the full system and that have caused issues in the past. I also advise to install cfclient and cflib inside a python environment as well.

**Reply by cuber-1** [reply]
I will give that a try! How exactly do I install the clients and libraries in a python environment?

**Reply by cuber-1** [reply]
Also to install python system wide, should I just reinstall it?

**Reply by cuber-1** [reply]
I installed python system wide by pressing the Install for all users button in the installer and then tried again. 

C:\Users\NAME>cfloader flash aideck_gap8_wifi_img_streamer_with_ap (1).bin deck-bcAI:gap8-fw -w radio://0/80/2M/E7E7E7E7E7
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\NAME\AppData\Roaming\Python\Python312\Scripts\cfloader.exe\__main__.py", line 7, in <module>
  File "C:\Users\NAME\AppData\Roaming\Python\Python312\site-packages\cfloader\__init__.py", line 116, in main
    [target, type] = t.split("-")
    ^^^^^^^^^^^^^^
ValueError: not enough values to unpack (expected 2, got 1)

Looks like a different issue now.

**Reply by cuber-1** [reply]
One more thing - when I connect to the Crazyflie, I see this as part of my console:

CPX: ESP32: I (435) SYS: Initialized
CPX: ESP32: W (435) ROUTER: Cannot route from GAP8 [0x00] to [0x00]
CPX: ESP32: I (435) DISCOVERY: Hostname is aideck-E78D6C
CPX: ESP32: I (435) DISCOVERY: MAC (STA) is c8:c9:a3:e7:8d:6c

I've seen the 1st, 3rd, and 4th message one other user's console, but I don't see the CPX: ESP32: W (435) ROUTER: Cannot route from GAP8 [0x00] to [0x00] on anyone else's.

**Reply by knmcguire** [reply]
hmm that is strange. Have you followed the initial instructions of updating the crazyflie firmware and the nina module, as mentioned here: https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#preparing-the-ai-deck-and-crazyflie-firmware ?

**Reply by cuber-1** [reply]
I figured out the issue - there were two problems: 1) I had downloaded the wifi streaming file multiple times which changed the name to something like aideck_gap8_wifi_img_streamer_with_ap (#).bin. As you can see, there is a space in the file name, which is what I believe triggered the issue. 2) The file I had was in the downloads section of my computer, as soon as I moved it to my user, it began to work fine. 

Now my next step is to update the bootloader as I the cfloader flashing process is stopping at 99%. I realized that I have a Rev. C1 AI deck, so I will be ordering a JTAG programmer. Could you confirm that this is the one I should buy?

https://www.digikey.com/en/products/detail/olimex-ltd/ARM-USB-TINY-H/3471388?utm_adgroup=&utm_source=google&utm_medium=cpc&utm_campaign=PMax%20Shopping_Product_Low%20ROAS%20Categories&utm_term=&utm_content=&utm_id=go_cmp-20243063506_adg-_ad-__dev-c_ext-_prd-3471388_sig-CjwKCAjwko21BhAPEiwAwfaQCN1ObM4peiJs_aMtPOv_6ZNydldOnSqhBbXgOZe8Vc1NQsWR3FU8sRoCfOcQAvD_BwE&gad_source=1&gclid=CjwKCAjwko21BhAPEiwAwfaQCN1ObM4peiJs_aMtPOv_6ZNydldOnSqhBbXgOZe8Vc1NQsWR3FU8sRoCfOcQAvD_BwE

Thanks for the continued assistance!

**Reply by knmcguire** [reply]
Yes that is the one! You can also get one of these as well: https://www.olimex.com/Products/ARM/JTAG/ARM-JTAG-20-10/, then it should be compatible with your lint cable. 

The debug adapter on our store will work for that too, but we don't sell the lint cable: https://store.bitcraze.io/collections/accessories/products/debug-adapter

**Reply by cuber-1** [reply]
So do I just need the product in the link I sent, or do I need to buy the first thing you linked as well?

**Reply by knmcguire** [reply]
You need both the programmer and the adapter. As you notice, the lint cable of the programmer won't fit on the 10 pin small header on the aideck, so you need something that goes from the big lint cable to a small one.

So the choices would be:
- Olimex programmer + the 20 - 10 jtag adapter 
- OR Olimex programmer + the bitcraze debug adapter + the 10 pin small lint cable

I would probably go for the first if digikey happens to also have the 20-10 jtag adapter

**Reply by cuber-1** [reply]
Ok, so would this be fine? https://www.digikey.com/en/products/detail/olimex-ltd/ARM-JTAG-20-10/3471401?utm_adgroup=&utm_source=google&utm_medium=cpc&utm_campaign=PMax%20Shopping_Product_Low%20ROAS%20Categories&utm_term=&utm_content=&utm_id=go_cmp-20243063506_adg-_ad-__dev-c_ext-_prd-3471401_sig-CjwKCAjw2dG1BhB4EiwA998cqDy1rqGzZ-XD_V9z6GMTEdy-22xCaelzxM9SNkjlM79fp8Uv_KMTgRoCeZMQAvD_BwE&gad_source=1&gclid=CjwKCAjw2dG1BhB4EiwA998cqDy1rqGzZ-XD_V9z6GMTEdy-22xCaelzxM9SNkjlM79fp8Uv_KMTgRoCeZMQAvD_BwE

**Reply by cuber-1** [reply]
Also, I'm not exactly sure how the 20 pin end of the adapter would connect to the 20 pin end of the JTAG programmer. Could you help me understand how they would be connected?

**Reply by knmcguire** [reply]
Yes that is the right adapter. the ARM-JTAG-20-10

The Olimex debugger link you sent from digikey (ARM-USB-TINY-H), you see that gray 20 pin cable in there? You take that out and you connect the adapter directly in the programmer like this:
![20240809_100822](https://github.com/user-attachments/assets/36dea7f1-d57b-4900-85af-62cebae1c63c)

**Reply by cuber-1** [reply]
And how should I connect the other end (the gray end) to my computer?

**Reply by gemenerik** [reply]
USB-B to A cable

**Reply by cuber-1** [reply]
I have received the JTAG programmer and adapter. I'm trying to run the given commands to update the bootloader but this is what I receive: 

cuber@NAME:/mnt/c/Users/NAMe/aideck-gap8-bootloader$ docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/;  make all image flash'
docker: unknown server OS: .
See 'docker run --help'.

Do I need to update it with a Linux OS?

**Reply by cuber-1** [reply]
So I'm using the virutal machine to update the bootloader now, and I'm facing a new issue:

I ran this command:

docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/;  make all image flash'

and I received:

make: *** No rule to make target 'all'. Stop.

**Reply by gemenerik** [reply]
Usually this means that you are not running the command from the correct directory, can you verify this?
